### PR TITLE
Chore: Refactored trax/tf_numpy/examples/mnist to make it work with tf.experimental.numpy

### PR DESCRIPTION
**Pull request by alfredoav**

### CODE REVIEW

File: trax/tf_numpy/examples/mnist/dataset.py [hunk 1]
```diff
@@ -13,77 +13,76 @@
...
+        elif split=='test':
+            mnist_images, mnist_labels = self.mnist_test_images, self.mnist_test_labels
+
+        len_dim_0 = mnist_images.shape[0]
+        mnist_images = np.reshape(mnist_images, (len_dim_0, 784))
+        mnist_labels = np.asarray(tf.one_hot(mnist_labels, 10))
+
+        mnist_images, mnist_labels = self.shuffle(mnist_images, mnist_labels)
+
+        index = 0
```
Comment (akshaym) [inline trax/tf_numpy/examples/mnist/dataset.py:71]: Is it possible to have the tfds.load(..., batch_size=self.batch_size)?

Alternatively, the loaded dataset can be batched directly? https://www.tensorflow.org/api_docs/python/tf/data/Dataset#batch

I see that this is yielding actual training data, it'd be nice to instead return an actual iterator on the underlying dataset.

File: trax/tf_numpy/examples/mnist/model.py [hunk 2]
```diff
@@ -19,119 +19,109 @@
...
-    Args:
-      x: 2-d array of size batch_size x image_size.
-
-    Returns:
-      A 2-d array of size batch_size x num_classes.
     """
 
-    def sigmoid(x):
-      return 1.0 / (1.0 + np.exp(-x))
+    def __init__(self, hidden_layers, input_size=784, num_classes=10, learning_rate=0.001):
```
Comment (akshaym) [inline trax/tf_numpy/examples/mnist/model.py:48]: learning_rate seems to be unused. I do prefer it here compared to on the train method, but its up to you if you want to keep as is.

File: trax/tf_numpy/examples/mnist/model.py [hunk 3]
```diff
@@ -19,119 +19,109 @@
...
+            be empty.
+          input_size: Length of the input array. The network receives the input
+            image as a flattened 1-d array. Defaults to 784(28*28), the default
+            image size for MNIST.
+          num_classes: The number of output classes. Defaults to 10.
+        """
+        hidden_layers = [input_size] + hidden_layers + [num_classes]
+        self.params = []
+        for i in range(len(hidden_layers) - 1):
+            self.params.append(np.random.randn(hidden_layers[i + 1], hidden_layers[i]))
```
Comment (akshaym) [inline trax/tf_numpy/examples/mnist/model.py:62]: How about a little more structure here?

Maybe the params could be a list of Namedtuples or even plain tuples? 

Layer = collections.namedtuple('Layer', ['weights', 'biases'])?

File: trax/tf_numpy/examples/mnist/model.py [hunk 4]
```diff
@@ -19,119 +19,109 @@
...
-    y_predicted = np.argmax(self.forward(x), axis=1)
-    return int(
-        np.sum(np.array(y_actual == y_predicted, copy=False, dtype=int32)))
+        out = self.sigmoid(np.dot(out, self.params[-2].T) + self.params[-1])
+
+        return out
+
+
+    def mean_squared_error(self, x, y):
+        y_out = self.forward(x)
```
Comment (akshaym) [inline trax/tf_numpy/examples/mnist/model.py:91]: Move the self.forward into the train loop (so "self.mean_squared_error(self.forward(x), y)" below), so its clear that mean_squared_error is just calculating the error and not the forward pass as well.

File: trax/tf_numpy/examples/mnist/model.py [hunk 5]
```diff
@@ -19,119 +19,109 @@
...
 
-  def evaluate(self, x, y):
-    """Returns the number of correct predictions.
+        out = tf.cast(x, tf.float32)
 
-    Args:
-      x: 2-d array of size batch_size x image_size.
-      y: 2-d array of size batch_size x num_classes.
+        for i in range(0, len(self.params)-2, 2):
+            out = np.tanh(np.dot(out, self.params[i].T) + self.params[i+1])
```
Comment (akshaym) [inline trax/tf_numpy/examples/mnist/model.py:83]: A few questions about this:
- The params seem to be transposed every time - it seems like we're doing initializing with randn(layer+1.size, layer.size). Can't we just switch that and you won't need the transpose?
- This was changed from sigmoid to tanh for the hidden layers - was there some particular reason for that?

File: trax/tf_numpy/examples/mnist/model.py [hunk 6]
```diff
@@ -19,119 +19,109 @@
...
-    total_len = len(new_weights_and_biases)
-    self.weights = new_weights_and_biases[:total_len // 2]
-    self.biases = new_weights_and_biases[total_len // 2:]
+        Returns:
+          A 2-d array of size batch_size x num_classes.
+        """
 
-  def evaluate(self, x, y):
-    """Returns the number of correct predictions.
+        out = tf.cast(x, tf.float32)
```
Comment (akshaym) [inline trax/tf_numpy/examples/mnist/model.py:80]: Why is this getting cast here?

If absolutely required, update the call to np.asarray.

File: trax/tf_numpy/examples/mnist/model.py [hunk 7]
```diff
@@ -19,119 +19,109 @@
...
-    return int(
-        np.sum(np.array(y_actual == y_predicted, copy=False, dtype=int32)))
+        out = self.sigmoid(np.dot(out, self.params[-2].T) + self.params[-1])
+
+        return out
+
+
+    def mean_squared_error(self, x, y):
+        y_out = self.forward(x)
+        loss = tf.math.reduce_sum((y - y_out)**2)
```
Comment (akshaym) [inline trax/tf_numpy/examples/mnist/model.py:92]: np.sum instead?

File: trax/tf_numpy/examples/mnist/train.py [hunk 8]
```diff
@@ -18,71 +18,57 @@
...
 import tensorflow.compat.v2 as tf
+from tensorflow.python.ops import numpy_ops as np
+import tensorflow_datasets as tfds
+
+from dataset import MNIST
+from model import Model
 
-from trax.tf_numpy.examples.mnist import dataset
-from trax.tf_numpy.examples.mnist import model as model_lib
+np.random.seed(0)
```
Comment (akshaym) [inline trax/tf_numpy/examples/mnist/train.py:30]: Move this into the "main" function.

File: trax/tf_numpy/examples/mnist/train.py [hunk 9]
```diff
@@ -18,71 +18,57 @@
...
+                print('iter {}: {} training accuracy, {} test accuracy.'.format(
+                    i+1,
+                    round(correct_train_predictions/mnist_dataset.train_len, 4),
+                    round(correct_val_predictions/mnist_dataset.val_len, 4)))
 
 
 if __name__ == '__main__':
   tf.compat.v1.enable_eager_execution()
-  app.run()
+  train(BATCH_SIZE, LEARNING_RATE, NUM_TRAINING_ITERS, VALIDATION_STEPS)
```
Comment (akshaym) [inline trax/tf_numpy/examples/mnist/train.py:74]: Let's change it back to app.run with a "def main". This is important for this to work reasonably internally.

File: trax/tf_numpy/examples/mnist/model.py (hunk 5 above)
Comment (alfredoav) [reply trax/tf_numpy/examples/mnist/model.py:83]: 1. Yes that makes more sense, I have now switched them.
2. Tanh tends to be a better midlayer to use because it is centered around 0 and also has larger derivatives. Also np does not include a native sigmoid so np.tanh is probably numerically stable than a self built sigmoid.

**Comment by alfredoav** [comment]
Dealing with this warning/error: W tensorflow/core/common_runtime/executor.cc:1086] [/device:CPU:0] Executor start aborting: Invalid argument: You must feed a value for placeholder tensor 'Placeholder/_3' with dtype int64 and shape [1]

**Comment by afrozenator** [comment]
@alfredoav - Many thanks for the contribution, but TF-NP is its own codebase now, after having gone out with TF 2.3.0 - I'm almost a 100% certain that this PR should go there.

@akshaym - I'm guessing this PR should go directly to the TF-NP codebase? Can you check with Ashish once? Thanks!

I will close this now, the TF-NP code will move out of Trax soon.

Many thanks though @alfredoav !
